include/ginkgo/core/distributed/index_map.hpp

```diff
@@ -0,0 +1,180 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_DISTRIBUTED_INDEX_MAP_HPP_
+#define GKO_PUBLIC_CORE_DISTRIBUTED_INDEX_MAP_HPP_
+
+
+#include <ginkgo/core/base/collection.hpp>
+#include <ginkgo/core/distributed/partition.hpp>
+
+
+namespace gko {
+namespace experimental {
+namespace distributed {
+
+
+/**
+ * \brief Index space classification for the locally stored indices.
+ */
+enum class index_space {
+    local,      //!< indices that are locally owned
+    non_local,  //!< indices that are owned by other processes
+    combined    //!< both local and non_local indices
+};
+
+
+/**
+ * \brief This class defines mappings between global and local indices.
+ *
+ * Given an index space $I = [0, \dots, N)$ that is partitioned into $P$
+ * disjoint subsets $I_k, k = 1, \dots, P$, this class defines for each
+ * subset an extended global index set $\hat{I}_k \supset I_K$. The extended
+ * index set contains the global indices owned by part $k$, as well as
+ * remote indices $R_k = \hat{I}_k \setminus I_k$, which are also accessed by
+ * part $k$, but owned by parts $l \neq k$.
+ * At the core, this class provides mappings from the global index space $I$
+ * into different local index spaces. The combined local index space
+ * (index_space::combined) is then defined as
+ * $[0, \dots, |\hat{I}_k|)$. Additionally, the combined index space can be
+ * separated into locally owned (index_space::local) and non-locally owned
+ * (index_space::non_local). The locally owned indices are defined as
+ * $[0, \dots, |I_k|)$, and the non-locally owned as $[0, \dots, |R_k|)$.
+ * With these index sets, the following mappings are defined:
+ * - $c_k : \hat{I}_k \mapsto [0, \dots, |\hat{I}_k|)$ which maps global indices
+ *   into the combined/full local index space,
+ * - $l_k: I_k \mapsto [0, \dots, |I_k|)$ which maps global indices into the
+ *   locally owned index space,
+ * - $r_k: R_k \mapsto [0, \dots, |R_k|)$ which maps global indices into the
+ *   non-locally owned index space.
+ * The required map can be selected by passing the appropriate type of an
+ * index_space.
+ *
+ * The index map for $I_k$ has no knowledge about any other index maps for
+ * $I_l, l \neq k$. In particular, any global index passed to the `get_local`
+ * map that is not part of the specified index space, will be mapped to an
+ * invalid_index.
+ *
+ * \tparam LocalIndexType  type for local indices
+ * \tparam GlobalIndexType  type for global indices
+ */
+template <typename LocalIndexType, typename GlobalIndexType = int64>
+struct index_map {
+    using part_type = Partition<LocalIndexType, GlobalIndexType>;
```

Comment:
I know partition is annoying to type out, but I believe we should avoid abbreviations if they themselves have another meaning in the same context (e.g. I used to refer to a specific rank's data in a partition as a part in a lot of my code, so this is surprising when skimming over)
```suggestion
    using partition_type = Partition<LocalIndexType, GlobalIndexType>;
```